c_src/erlang_jq_nif.c

```diff
@@ -253,9 +269,71 @@ static ERL_NIF_TERM make_ok_return(ErlNifEnv* env, ERL_NIF_TERM result) {
     return enif_make_tuple2(env, enif_make_atom(env, "ok"), result);
 }
 
+typedef struct {
+    ErlNifMutex* lock;
+    volatile bool consumed_by_process_json_nif;
+    jq_state* state;
+} JQCancelableStateHolderResource;
+
+static ERL_NIF_TERM
+create_jq_resource_nif(
+        ErlNifEnv* env,
+        int argc,
+        const ERL_NIF_TERM argv[]) {
+    module_private_data* module_data = enif_priv_data(env);
+    JQCancelableStateHolderResource* obj =
+        enif_alloc_resource(module_data->jq_state_holder_resource_type,
+                            sizeof(JQCancelableStateHolderResource));
+    obj->state = NULL;
+    static char* lock_name = "jq_nif_resource_lock";
+    obj->lock = enif_mutex_create(lock_name);
+    obj->consumed_by_process_json_nif = false;
+    // Start by locking the lock to make sure jq_state is set before timeout
+    // The lock needs to be unlocked in process_json_with_jq_resource_nif after
+    // the state obj->state field has been set or after an error has been
+    // detected.
+    enif_mutex_lock(obj->lock);
```

Comment:
Am I right if one process call `create_jq_resource` then `cancel_jq_resource` the scheduler will be dead locked?

Reply:
No, `cancel_jq_resource_nif` will return without acquiring any lock unless the field `->consumed_by_process_json_nif` in the resource is set to `true`. The field `->consumed_by_process_json_nif` is only set to `true` in `process_json_nif` and in all places where `->consumed_by_process_json_nif` is set to `true` the NIF resource's lock is unlocked on the next line so a deadlock of the scheduler should never happen. This is quite tricky code so I would appreciate if you can help me confirm this. Don't hesitate to ask if there is something in the code that you don't understand!